File: Dynet.py
from __future__ import annotations

from copy import deepcopy
from random import randint, choice, random, uniform as randfloat
from typing import List, Callable

from numpy import tanh
from numpy import exp

import time

# //  _____                _           _   ______
# // /  __ \              | |         | |  | ___ \
# // | /  \/_ __ ___  __ _| |_ ___  __| |  | |_/ /_   _
# // | |   | '__/ _ \/ _` | __/ _ \/ _` |  | ___ \ | | |
# // | \__/\ | |  __/ (_| | ||  __/ (_| |  | |_/ / |_| |
# //  \____/_|  \___|\__,_|\__\___|\__,_|  \____/ \__, |
# //                                               __/ |
# //                                              |___/
# //   ___        _ _     ______     _       _
# //  / _ \      (_) |    | ___ \   | |     | |
# // / /_\ \_ __  _| | __ | |_/ /_ _| |_ ___| |
# // |  _  | '_ \| | |/ / |  __/ _` | __/ _ \ |
# // | | | | | | | |   <  | | | (_| | ||  __/ |
# // \_| |_/_| |_|_|_|\_\ \_|  \__,_|\__\___|_|


IN = 0
HIDDEN = 1
OUT = 2

SIGMOID = 0
TANH = 1

def sigmoid(x):
    return 1/(1 + exp(-x))


class Connection:
    """
    Represents a connection between two neurons
    """
    def __init__(self, layerFrom: int, indexFrom: int, weight: float):
        """
        Creates a connection.

        :param layerFrom: What layer (inout, hidden or output) either 0, 1, 2
        :param indexFrom: Index of the neuron in the layer
        :param weight: Weight of connection
        """
        self.layerFrom = layerFrom
        self.indexFrom = indexFrom
        self.weight = weight


class Neuron:
    """
    Represents a single Neuron
    """
    def __init__(self):
        """
        Creates a single neuron
        """
        self.connections = []
        self.outGoing = 0
        self.outGoingConnections = []
        self.value = 0
        self.bias = 0

    def addConnection(self, layerFrom: int, indexFrom: int, weight: float):
        """
        Add a connection.

        :param layerFrom: What layer (inout, hidden or output) either 0, 1, 2
        :param indexFrom: Index of the neuron in the layer
        :param weight: Weight of connection
        """
        self.connections.append(Connection(layerFrom, indexFrom, weight))


class Dynet:
    """
    Represents an entire Dynet
    """
    def __init__(self, inputs, outputs, hiddens=1, activation=SIGMOID):
        """
        Create an entire Dynet

        :param inputs: The number of input neurons
        :param hiddens: The number of hidden neurons
        :param outputs: The number of output neurons
        """
        h = hiddens
        if h <= 0:
            h = 1
        self.inputs = [Neuron() for _ in range(inputs)]
        self.hiddens = [Neuron() for _ in range(h)]
        self.outputs = [Neuron() for _ in range(outputs)]
        self.weightRange = 1
        self.activation = activation

    def activate(self, x):
        if self.activation == SIGMOID:
            return sigmoid(x)
        else:
            return tanh(x)

    def derivative_activate(self, out):
        if self.activation == SIGMOID:
            return out * (1 - out)

    def addRandomInputToHiddenConnection(self):
        """
        Adds a random input to the hidden layer
        """
        randomInputIndex = randint(0, len(self.inputs) - 1)
        randomHiddenIndex = randint(0, len(self.hiddens) - 1)
        self.hiddens[randomHiddenIndex].addConnection(
            IN, randomInputIndex, randfloat(-self.weightRange, self.weightRange)
        )
        self.inputs[randomInputIndex].outGoing += 1

    def addRandomInputToOutputConnection(self):
        """
        Adds a random output to the hidden layer
        """
        randomInputIndex = randint(0, len(self.inputs) - 1)
        randomOutputIndex = randint(0, len(self.outputs) - 1)
        self.outputs[randomOutputIndex].addConnection(
            IN, randomInputIndex, randfloat(-self.weightRange, self.weightRange)
        )
        self.inputs[randomInputIndex].outGoing += 1
        self.inputs[randomInputIndex].outGoingConnections.append([OUT, randomOutputIndex, len(self.outputs[randomOutputIndex].connections)-1])

    def addRandomHiddenToHiddenConnection(self):
        """
        Adds a random hidden to the hidden layer
        """
        randomHiddenIndex = randint(0, len(self.hiddens) - 1)
        randomHiddenIndex2 = randint(0, len(self.hiddens) - 1)
        self.hiddens[randomHiddenIndex2].addConnection(
            HIDDEN, randomHiddenIndex, randfloat(-self.weightRange, self.weightRange)
        )
        self.hiddens[randomHiddenIndex].outGoing += 1
        self.hiddens[randomHiddenIndex].outGoingConnections.append([HIDDEN, randomHiddenIndex2, len(self.hiddens[randomHiddenIndex2].connections)-1])

    def addRandomHiddenToOutputConnection(self):
        """
        Adds a random hidden to the output layer
        """
        randomHiddenIndex = randint(0, len(self.hiddens) - 1)
        randomOutputIndex = randint(0, len(self.outputs) - 1)
        self.outputs[randomOutputIndex].addConnection(
            HIDDEN, randomHiddenIndex, randfloat(-self.weightRange, self.weightRange)
        )
        self.hiddens[randomHiddenIndex].outGoing += 1
        self.hiddens[randomHiddenIndex].outGoingConnections.append([OUT, randomOutputIndex, len(self.outputs[randomOutputIndex].connections)-1])

    def addRandomConnection(self):
        """
        Add a random connection somewhere
        """
        funcs = [self.addRandomInputToHiddenConnection,
                 self.addRandomInputToOutputConnection,
                 self.addRandomHiddenToHiddenConnection,
                 self.addRandomHiddenToOutputConnection]
        try:
            choice(funcs)()
        except IndexError:
            pass

    def removeRandomConnectionHiddens(self):

        randomHiddenIndex = randint(0, len(self.hiddens) - 1)
        if len(self.hiddens[randomHiddenIndex].connections) > 0:
            randomConnectionIndex = randint(0, len(self.hiddens[randomHiddenIndex].connections)-1)
            self.hiddens[randomHiddenIndex].connections.pop(randomConnectionIndex)
    def removeRandomConnectionOutputs(self):
        randomOutputIndex = randint(0, len(self.outputs) - 1)
        if len(self.outputs[randomOutputIndex].connections) > 0:
            randomConnectionIndex = randint(0, len(self.outputs[randomOutputIndex].connections)-1)
            self.outputs[randomOutputIndex].connections.pop(randomConnectionIndex)


    def removeRandomConnection(self):
        if random() < 0.5:
            self.removeRandomConnectionHiddens()
        else:
            self.removeRandomConnectionOutputs()

    def mutateRandomConnection(self):
        if random() < 0.5:
            randomHiddenIndex = randint(0, len(self.hiddens) - 1)
            if len(self.hiddens[randomHiddenIndex].connections) > 0:
                rConn = randint(0, len(self.hiddens[randomHiddenIndex].connections)-1)
                self.hiddens[randomHiddenIndex].connections[rConn].weight+=randfloat(-0.1, 0.1)
        else:
            randomOutputIndex = randint(0, len(self.outputs) - 1)
            if len(self.outputs[randomOutputIndex].connections) > 0:
                rConn = randint(0, len(self.outputs[randomOutputIndex].connections) - 1)
                self.outputs[randomOutputIndex].connections[rConn].weight += randfloat(-0.1, 0.1)

    def weightedSumHiddens(self):
        """
        Calculates the weighted sum of each hidden neuron and does stuff
        """
        for neuron in self.hiddens:
            neuron.value = 0
            if neuron.outGoing == 0:
                continue
            for conn in neuron.connections:
                if conn.layerFrom == IN:
                    neuron.value += conn.weight * self.inputs[conn.indexFrom].value
                elif conn.layerFrom == HIDDEN:
                    neuron.value += conn.weight * self.hiddens[conn.indexFrom].value
            neuron.value += neuron.bias
            neuron.value = self.activate(neuron.value)

    def weightedSumOutputs(self):
        """
        Calculates the weighted sum of each output neuron and does stuff
        """
        for neuron in self.outputs:
            neuron.value = 0
            for conn in neuron.connections:
                if conn.layerFrom == HIDDEN:
                    neuron.value += conn.weight * self.hiddens[conn.indexFrom].value
                elif conn.layerFrom == IN:
                    neuron.value += conn.weight * self.inputs[conn.indexFrom].value
            neuron.value = self.activate(neuron.value)


    def mutateBias(self):
        if random() < 0.5:
            r = randint(0, len(self.outputs)-1)
            self.outputs[r].bias += randfloat(-0.1, 0.1)
        elif len(self.hiddens) > 0:
            r = randint(0, len(self.hiddens) - 1)
            self.hiddens[r].bias += randfloat(-0.1, 0.1)

    def mutate(self, rate: float, repeats: int, modifyHiddens=True):
        """
        Mutate stuff

        :param rate: Mutation rate
        :param repeats: How many times to repeat it
        """
        for _ in range(repeats):
            if random() < rate / 2:
                self.addRandomConnection()
            if random() < rate:
                self.mutateRandomConnection()
            if random() < rate / 20 and modifyHiddens:
                self.hiddens.append(Neuron())
            if random() < rate / 2:
                self.removeRandomConnection()
            if random() < rate:
                self.mutateBias()

    def fullyConnect(self):
        #Connect all Inputs to all hiddens
        for index, _ in enumerate(self.inputs):
            for hidden in self.hiddens:
                hidden.addConnection(IN, index, randfloat(-self.weightRange, self.weightRange))
                self.inputs[index].outGoing += 1

        #Connect all hiddens to each other
        for index, _ in enumerate(self.hiddens):
            for i, hidden in enumerate(self.hiddens):
                if i == index:
                    continue
                hidden.addConnection(HIDDEN, index, randfloat(-self.weightRange, self.weightRange))
                self.hiddens[index].outGoing += 1

        #Connect all hiddens to outputs:
        for index, _ in enumerate(self.hiddens):
            for output in self.outputs:
                output.addConnection(HIDDEN, index, randfloat(-self.weightRange, self.weightRange))
                self.hiddens[index].outGoing+=1

    def backpropagate(self, expected):
        outputs = []
        outputErrors = []
        hiddenOutputError = []
        hiddenHiddenError = []
        inputHiddenError = []
        ##Calculate outputs errors
        for i in self.outputs:
            outputs.append(i.value)
        for index, i in enumerate(outputs):
            outputErrors.append((i-expected[index]) ** 2)

        ##Calculate hidden to output error
        for index, output in enumerate(self.outputs):
            totalError = 0
            for conIndex, conn in enumerate(output.connections):
                totalError += (conn.weight * outputErrors[index]) * self.derivative_activate(self.hiddens[conn.indexFrom].value)
            hiddenOutputError.append(totalError)

        # for index, hidden in enumerate(self.hiddens):
        #     totalError = 0
        #     for conIndex, conn in enumerate(hidden.connections):
        #         if (conn.layerFrom != HIDDEN):
        #             continue
        #         totalError += (conn.weight * hiddenOutputError[index]) * self.derivative_activate(self.hiddens[conn.indexFrom].value)
        #     hiddenHiddenError.append(totalError)

        print(outputErrors, hiddenOutputError)



    def feedForward(self, ins: List[int], printTime=False) -> List[float]:
        """
        Feed forward something

        :param ins: A list of integers
        """
        t = time.perf_counter()
        outputs = []
        for index, i in enumerate(ins):
            self.inputs[index].value = i
        self.weightedSumHiddens()
        self.weightedSumOutputs()
        for i in self.outputs:
            outputs.append(i.value)

        if printTime:
            print(time.perf_counter()-t)
        return outputs

    def printNetwork(self, printFunc: Callable = print):
        """
        Print the network.

        :param printFunc: The print function to use. Defaults to the default
         print function.
        """
        printFunc("Network: ")
        totalCons = 0
        for i in self.inputs:
            printFunc(f"{i.value:.2f} ", end="")
            totalCons += len(i.connections)
        printFunc("")
        for i in self.hiddens:
            printFunc(f"{i.value:.2f} ", end="")
            totalCons += len(i.connections)
        printFunc("")
        for i in self.outputs:
            printFunc(f"{i.value:.2f} ", end="")
            totalCons += len(i.connections)
        printFunc("")
        printFunc(f"Connection count: {totalCons}")

    def copy(self) -> Dynet:
        """
        Returns a copy of oneself, for mutating
        """
        return deepcopy(self)

File: JS_PORT/Dynet.js
//  _____                _           _   ______
// /  __ \              | |         | |  | ___ \
// | /  \/_ __ ___  __ _| |_ ___  __| |  | |_/ /_   _
// | |   | '__/ _ \/ _` | __/ _ \/ _` |  | ___ \ | | |
// | \__/\ | |  __/ (_| | ||  __/ (_| |  | |_/ / |_| |
//  \____/_|  \___|\__,_|\__\___|\__,_|  \____/ \__, |
//                                               __/ |
//                                              |___/
//   ___        _ _     ______     _       _
//  / _ \      (_) |    | ___ \   | |     | |
// / /_\ \_ __  _| | __ | |_/ /_ _| |_ ___| |
// |  _  | '_ \| | |/ / |  __/ _` | __/ _ \ |
// | | | | | | | |   <  | | | (_| | ||  __/ |
// \_| |_/_| |_|_|_|\_\ \_|  \__,_|\__\___|_|

// Constants
const IN = 0;
const HIDDEN = 1;
const OUT = 2;

// Available activation functions
const SIGMOID = 0;
const TANH = 1;

function sigmoid(x) {
  return 1 / (1 + exp(-x));
}

function randf(lo, hi) {
  return Math.random() * (hi + 1) + lo;
}

function randint(lo, hi) {
  return Math.floor(randf(lo, hi));
}

// Clones an object
function clone(target, map = new WeakMap()) {
  if (typeof target === "object") {
    let cloneTarget = Array.isArray(target) ? [] : {};
    if (map.get(target)) {
      return map.get(target);
    }
    map.set(target, cloneTarget);
    for (const key in target) {
      cloneTarget[key] = clone(target[key], map);
    }
    return cloneTarget;
  } else {
    return target;
  }
};

// Represents a connection between layers
class Connection {
  constructor(layerFrom, indexFrom, weight) {
    this.layerFrom = layerFrom;
    this.indexFrom = indexFrom;
    this.weight = weight;
  }

  copy() {
    let c = new Connection(this.layerFrom, this.indexFrom, this.weight);
    return c;
  }

}

// Represents a single neuron
class Neuron {
  constructor() {
    this.connections = [];
    this.outGoing = 0;
    this.outGoingConnections = [];
    this.value = 0;
    this.bias = 0;
  }

  addConnection(layerFrom, indexFrom, weight) {
    this.connections.push(new Connection(layerFrom, indexFrom, weight));
  }

  copy() {
    let c = new Neuron();
    for (let con of this.connections) {
      c.connections.push(con.copy());
    }
    return c;
  }
}

// Represents an entire Dynet network
class Dynet {
  constructor(inputs, outputs, hiddens=1, activation=SIGMOID) {
    // Must be greater then or equal to 1
    if (hiddens < 1) {
      hiddens = 1;
    }
    this.activation = activation;
    this.weightRange = 1;
    // Create the neurons
    this.inputs = [];
    this.outputs = [];
    this.hiddens = [];
    for (let i = 0; i < inputs; i ++) {
      this.inputs.push(new Neuron());
    }
    for (let i = 0; i < outputs; i ++) {
      this.outputs.push(new Neuron());
    }
    for (let i = 0; i < hiddens; i ++) {
      this.hiddens.push(new Neuron());
    }
  }

  activate(x) {
    // Run the activation function that was specified in the constructor
    if (this.activation == SIGMOID) {
      return sigmoid(x);
    }
    else if (this.activation == TANH) {
      return Math.tanh(x);
    }
  }

  addRandomInputToHiddenConnection() {
    let randomInputIndex = randint(0, this.inputs.length - 1);
    let randomHiddenIndex = randint(0, this.hiddens.length - 1);
    this.hiddens[randomHiddenIndex].addConnection(
      IN, randomInputIndex, randf(-this.weightRange, this.weightRange)
    );
    this.inputs[randomInputIndex].outGoing += 1;
  }

  addRandomInputToOutputConnection() {
    let randomInputIndex = randint(0, this.inputs.length - 1);
    let randomOutputIndex = randint(0, this.outputs.length - 1);

    this.outputs[randomOutputIndex].addConnection(
      IN, randomInputIndex, randf(-this.weightRange, this.weightRange)
    );
    this.inputs[randomInputIndex].outGoing += 1;
  }

  addRandomHiddenToOutputConnection() {
    let randomHiddenIndex = randint(0, this.hiddens.length - 1);
    let randomOutputIndex = randint(0, this.outputs.length - 1);
    this.outputs[randomOutputIndex].addConnection(
      HIDDEN, randomHiddenIndex, randf(-this.weightRange, this.weightRange)
    );
    this.hiddens[randomHiddenIndex].outGoing += 1;
  }

  addRandomHiddenToHiddenConnection() {
    let randomHiddenIndex = randint(0, this.hiddens.length - 1);
    let randomHidden2Index = randint(0, this.hiddens.length - 1);
    this.hiddens[randomHidden2Index].addConnection(
        HIDDEN, randomHiddenIndex, randf(-this.weightRange, this.weightRange)
    )
    this.hiddens[randomHiddenIndex].outGoing += 1;
  }

  addRandomConnection() {
    let r = randint(0, 4)
    if (r == 0){
      this.addRandomHiddenToHiddenConnection();
    }
    else if (r == 1){
      this.addRandomHiddenToOutputConnection();
    }
    else if (r == 2){
      this.addRandomInputToHiddenConnection();
    }
    else{
      this.addRandomInputToOutputConnection();
    }
  }

  removeRandomConnectionHiddens() {
    let randomHiddenIndex = randint(0, this.hiddens.length-1);
    if (this.hiddens[randomHiddenIndex].connections.length > 0) {
      let randomConnectionIndex = randint(0, this.hiddens[randomHiddenIndex].connections.length-1);
      this.hiddens[randomHiddenIndex].connections.pop(randomConnectionIndex);
    }
  }

  removeRandomConnectionOutputs() {
    let randomOutputIndex = randint(0, this.outputs.length-1);
    if (this.outputs[randomOutputIndex].connections.length > 0) {
      let randomConnectionIndex = randint(0, this.outputs[randomOutputIndex].connections.length-1);
      this.outputs[randomOutputIndex].connections.pop(randomConnectionIndex);
    }
  }

  removeRandomConnection() {
    // Remove a random connection
    if (random() < 0.5) {
      this.removeRandomConnectionHiddens();
    } else {
      this.removeRandomConnectionOutputs();
    }
  }

  constrainAllHiddenConnections(){
    for (let neuron of this.hiddens){
      for (let conn of neuron.connections){
        if (conn.weight < -this.weightRange){
          conn.weight = -this.weightRange
        }
        if (conn.weight > this.weightRange){
          conn.weight = this.weightRange
        }
      }
    }
  }

  constrainAllOutputConnections() {
    for (let neuron of this.outputs) {
      for (let conn of neuron.connections) {
        if (conn.weight < -this.weightRange) {
          conn.weight = -this.weightRange;
        }
        if (conn.weight > this.weightRange) {
          conn.weight = this.weightRange;
        }
      }
    }
  }

  mutateRandomConnection() {
    // Mutate a random connection
    if (random() < 0.5) {
      let randomHiddenIndex = randint(0, this.hiddens.length-1);
      if (this.hiddens[randomHiddenIndex].connections.length > 0) {
        let rConn = randint(0, this.hiddens[randomHiddenIndex].connections.length-1);
        this.hiddens[randomHiddenIndex].connections[rConn].weight += randf(-0.1, 0.1);

      }
    } else {
      let randomOutputIndex = randint(0, this.outputs.length-1);
      if (this.outputs[randomOutputIndex].connections.length > 0) {
        let rConn = randint(0, this.outputs[randomOutputIndex].connections.length-1);
        this.outputs[randomOutputIndex].connections[rConn].weight += randf(-0.1, 0.1);
      }
    }
  }

  weightedSumHiddens() {
    for (let neuron of this.hiddens) {
      neuron.value = 0;
      if (neuron.outGoing == 0) {
        //continue;
      }
      for (let conn of neuron.connections) {
        if (conn.layerFrom == IN) {
          neuron.value += conn.weight * this.inputs[conn.indexFrom].value;
        } else if (conn.layerFrom == HIDDEN) {
          neuron.value += conn.weight * this.hiddens[conn.indexFrom].value;
        }
      }
      neuron.value += neuron.bias;
      neuron.value = this.activate(neuron.value);
    }
  }

  weightedSumOutputs() {
    for (let neuron of this.outputs) {
        neuron.value = 0;
        for (let conn of neuron.connections) {
          if (conn.layerFrom == HIDDEN) {
            neuron.value += conn.weight * this.hiddens[conn.indexFrom].value;
          } else if (conn.layerFrom == IN) {
            neuron.value += conn.weight * this.inputs[conn.indexFrom].value;
          }
        }
        neuron.value = this.activate(neuron.value);
    }
  }
  fullyConnect(){
    for (let neuron in this.outputs){
      for (let n2 in this.hiddens){
        this.outputs[neuron].addConnection(HIDDEN, n2, randf(-this.weightRange, this.weightRange))
      }
    }
    for (let neuron in this.hiddens){
      for (let n2 in this.hiddens){
        this.hiddens[neuron].addConnection(HIDDEN, n2, randf(-this.weightRange, this.weightRange))
      }
    }
    for (let neuron in this.hiddens){
      for (let n2 in this.inputs){
        this.hiddens[neuron].addConnection(IN, n2, randf(-this.weightRange, this.weightRange))
      }
    }
  }
  mutateBias() {
    if (random() < 0.5) {
      let r = randint(0, this.outputs.length-1);
      this.outputs[r].bias += randf(-0.1, 0.1);
    } else if (this.hiddens.length > 0) {
      let r = randint(0, this.hiddens.length-1);
      this.hiddens[r].bias += randf(-0.1, 0.1);
    }
  }

  mutate(rate, repeats, modifyHiddens=true) {
    for (let i = 0; i < repeats; i ++) {
      if (random() < rate / 5) {
        this.addRandomConnection();
      }
      if (random() < rate) {
        this.mutateRandomConnection();
      }
      if (random() < rate / 20 && modifyHiddens) {
        this.hiddens.push(new Neuron());
      }
      if (random() < rate / 5) {
        this.removeRandomConnection();
      }
      if (random() < rate) {
        this.mutateBias();
      }
      this.constrainAllHiddenConnections()
      this.constrainAllOutputConnections()
    }
  }

  feedForward(ins) {
    let outputs = [];
    for (let i = 0; i < this.inputs.length; i++) {
      this.inputs[i].value = ins[i];
    }
    this.weightedSumHiddens();
    this.weightedSumOutputs();
    for (let i = 0; i < this.outputs.length; i++) {
      outputs.push(this.outputs[i].value);
    }
    return outputs;
  }

  printNetwork() {
    console.log("Network: ");
    let totalCons = 0;
    for (let i of this.inputs) {
      totalCons += i.connections.length;
    }
    for (let i of this.hiddens) {
      totalCons += i.connections.length;
    }
    for (let i of this.outputs) {
      totalCons += i.connections.length;
    }
    console.log("Connection count: " + totalCons);
  }

  copy() {
    let c = new Dynet(this.inputs.length, this.outputs.length, this.hiddens.length, this.activation);
    for (let i in this.inputs) {
      c.inputs[i] = this.inputs[i].copy();
    }
    for (let i in this.hiddens) {
      c.hiddens[i] = this.hiddens[i].copy();
    }
    for (let i in this.outputs) {
      c.outputs[i] = this.outputs[i].copy();
    }
    return c;
  }
}
